kuma/wiki/apps.py

```diff
@@ -1,48 +1,49 @@
 from django.apps import AppConfig
 from django.conf import settings
-from django.core.exceptions import ObjectDoesNotExist
 from django.db.models import signals
 from django.utils.translation import ugettext_lazy as _
 
 from elasticsearch_dsl.connections import connections as es_connections
 
-from .jobs import (DocumentContributorsJob, DocumentZoneStackJob,
+from .models import Document, DocumentZone
+from .jobs import (DocumentContributorsJob, DocumentNearestZoneJob,
                    DocumentZoneURLRemapsJob, DocumentCodeSampleJob)
 from .signals import render_done
 
 
-def invalidate_zone_stack_cache(document, async=False):
+def invalidate_nearest_zone_cache(document_pk, async=False):
     """
-    Reset the cache for the zone stack for all of the documents
-    in the document tree branch.
+    Reset the nearest-zone cache for this document and its descendants.
     """
-    pks = [document.pk] + [parent.pk for parent in
-                           document.get_topic_parents()]
-    job = DocumentZoneStackJob()
-    if async:
-        invalidator = job.invalidate
-    else:
-        invalidator = job.refresh
-    for pk in pks:
-        invalidator(pk)
-
-
-def invalidate_zone_urls_cache(document, async=False):
+    job = DocumentNearestZoneJob()
+    do_invalidate = job.invalidate if async else job.refresh
+
+    def invalidate(pk):
+        do_invalidate(pk)
+        # Since the descendants of this document search upwards for their
+        # nearest zone, recursively invalidate their caches. Note that the
+        # branches of this tree of decendants can stop at (and exclude) any
+        # descendants that have their own zones.
+        children = (Document.objects
+                            .filter(parent_topic=pk)
+                            .values_list('pk', 'zone'))
```

Comment:
Nitpick: Using ``zone__id`` (double underscore) will just load the ID instead of a full ``DocumentZone`` instance.